src/Features/Core/Portable/UseThrowExpression/AbstractUseThrowExpressionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,286 @@
+    /// expressions as well.
+    /// </summary>
+    internal abstract class AbstractUseThrowExpressionDiagnosticAnalyzer : DiagnosticAnalyzer, IBuiltInAnalyzer
+    {
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(FeaturesResources.Use_throw_expression), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(FeaturesResources.Use_throw_expression), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+
+        private static DiagnosticDescriptor s_descriptor = new DiagnosticDescriptor(
+                IDEDiagnosticIds.UseThrowExpressionDiagnosticId,
+                s_localizableTitle,
+                s_localizableMessage,
+                DiagnosticCategory.Style,
+                DiagnosticSeverity.Hidden,
+                isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics 
+            => ImmutableArray.Create(s_descriptor);
+
+        public DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        public bool OpenFileOnly(Workspace workspace) => false;
+
+        private static MethodInfo s_registerOperationActionInfo =
+            typeof(AnalysisContext).GetTypeInfo().GetDeclaredMethod("RegisterOperationActionImmutableArrayInternal");
+
+        private static MethodInfo s_getOperationInfo =
+            typeof(SemanticModel).GetTypeInfo().GetDeclaredMethod("GetOperationInternal");
+
+        protected abstract bool IsSupported(ParseOptions options);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            s_registerOperationActionInfo.Invoke(context, new object[]
+            {
+                new Action<OperationAnalysisContext>(AnalyzeOperation),
+                ImmutableArray.Create(OperationKind.ThrowStatement)
+            });
+        }
+
+        private void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            if (!IsSupported(context.Operation.Syntax.SyntaxTree.Options))
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+
+            var throwOperation = (IThrowStatement)context.Operation;
+            var throwStatement = throwOperation.Syntax;
+
+            var optionSet = context.Options.GetOptionSet();
+            var option = optionSet.GetOption(CodeStyleOptions.PreferThrowExpression, throwStatement.Language);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var compilation = context.Compilation;
+            var semanticModel = compilation.GetSemanticModel(throwStatement.SyntaxTree);
+
+            var ifOperation = GetContainingIfOperation(
```

Comment:
This is very clever, but you shouldn't have to do it. We need to add something to the IOperation API to fetch parents.